sbt-scala-native/src/main/scala/scala/scalanative/sbtplugin/ScalaNativePluginInternal.scala

```diff
@@ -109,28 +110,75 @@ object ScalaNativePluginInternal {
         .withDump(nativeDump.value)
     },
     nativeLink := {
+      val classpath = fullClasspath.value.map(_.data.toPath)
       val outpath = (nativeLink / artifactPath).value
-      val config = {
-        val mainClass = selectMainClass.value.getOrElse {
-          throw new MessageOnlyException("No main class detected.")
+
+      def makeBuild(): HashFileInfo = {
+        val config = {
+          val mainClass = selectMainClass.value.getOrElse {
+            throw new MessageOnlyException("No main class detected.")
+          }
+
+          val maincls = mainClass + "$"
+          val cwd = nativeWorkdir.value.toPath
+
+          val logger = streams.value.log.toLogger
+          build.Config.empty
+            .withLogger(logger)
+            .withMainClass(maincls)
+            .withClassPath(classpath)
+            .withWorkdir(cwd)
+            .withCompilerConfig(nativeConfig.value)
+        }
+
+        interceptBuildException {
+          Build.build(config, outpath.toPath)(sharedScope)
         }
-        val classpath = fullClasspath.value.map(_.data.toPath)
-        val maincls = mainClass + "$"
-        val cwd = nativeWorkdir.value.toPath
-
-        val logger = streams.value.log.toLogger
-        build.Config.empty
-          .withLogger(logger)
-          .withMainClass(maincls)
-          .withClassPath(classpath)
-          .withWorkdir(cwd)
-          .withCompilerConfig(nativeConfig.value)
+
+        FileInfo.hash(outpath)
```

Comment:
I think it would be great if we could move hashing the output path outside this method to follow the single-responsibility principle. Also since we have the method `buildIfChanged` maybe let's rename it to simply `build`?  
I'm also wondering wouldn't last modified info be enough for our use case, instead of hashing its binary content.

Reply:
Initially last modified was used, but when trying to fix some other (unrelated) issue, I found this: https://www.boost.org/doc/libs/1_32_0/libs/filesystem/doc/operations.htm#last_write_time , specifically the warning under last_write_time. Apparently the last modified value can have quite a low resolution depending on the os and filesystem, although I have not found any more info than this. I simply thought of hashing as a safer choice.